Add workflow for EDS notebook submission

abnerbog opened the pull request:
This PR (copy of https://github.com/abnerbog/from-scripts-to-pipelines-workshop/pull/1) adds the `targets` workflow for retrieving, processing and visualizing Hurricane Harvey best track NHC data, stage data from USGS and flood data from NWS on a small number of sites in Texas and Louisiana with uniquely shaped hydrographs.

The primary outputs are:
- `04_visualize/out/leaflet_map.html` -- leaflet map of hurricane track over time along with USGS gage locations
![image](https://github.com/user-attachments/assets/da115a8e-ed98-4bb2-b2bd-4c1dbcfcdecb)
- `04_visualize/out/hydrograph_plot.svg` -- plots of stage data over time along with minor flood stage for reference
![image](https://github.com/user-attachments/assets/d1d9c050-e94d-477d-9ed0-f3d7ea4d9ec7)

**Lindsay, here are the things that I think are worth focusing on reviewing in this PR (along with anything else you notice):**
- Can you run the pipeline?
  - Does it run if you add/change/modify one of the USGS gage sites? (must be in TX or LA)
- Are the primary visualization outputs appreciable? Any major changes you might recommend here?
- Are the intermediate .csv files appreciable? The csv outputs are in:
  - `01_fetch/out/nwis_{site_no}_data.csv`
  - `02_read/out/nwis_site_info.csv`
  - `03_process/out/harvey_tracking.csv`

abnerbog committed:
Add gitignore

abnerbog committed:
Rename and add phases

abnerbog committed:
Update recipe file

abnerbog committed:
Add fetch/download step with ref scripts

abnerbog committed:
Add placeholder in tmp folder in fetch/download

abnerbog committed:
Add ref scripts scripts for read step

abnerbog committed:
Add process step with ref scripts

abnerbog committed:
Add viz step with ref scripts

abnerbog committed:
Update gitignore

abnerbog committed:
Add DAG of initial pipeline

abnerbog committed:
Replace pipeline viz from html to png

lindsayplatt commented:
@abnerbog I have this logged for a review at the end of this week. Can you check out why those conflicts exist and resolve them first?

abnerbog committed:
Fix merge conflicts

lindsayplatt reviewed (commented):
Hey Abner! Despite my many comments, this is looking great. And it is very exciting to see that you were able to take that code from a much older pipeline and turn it into a `targets` one!

Major things:
1. I am not able to build this alll the way through. I am getting an error on `p3_harvey_pts` and left some details there. Please take a more detailed look at the build.
2. Once this is working, I suggest more documentation! Having the Roxygen style function documentation (could plug fxns into ChatGPT to get a significant start on this) would be helpful. Plus, more in-line comments in the target recipes to help guide users.

Really good start here!

abnerbog committed:
Updating inputs to p3_harvey_pts target

abnerbog committed:
Autoidenting

abnerbog committed:
Removing unused code

abnerbog committed:
Added target for intermediate csv containing nwis site data

abnerbog commented on 01_fetch.R:
```diff
@@ -2,28 +2,39 @@
 p1 <- list(
+  tar_target(
+    p1_site_data,
+    download_nwis_stage_data(sites=harvey_sites,
+                                 start_date=start_date,
+                                 end_date=end_date)
```
Didn't know about the autoindent command, thanks!

abnerbog commented on 01_fetch/src/download_nwis_stage_data.R:
```diff
@@ -0,0 +1,37 @@
+    file_out <- file.path("01_fetch/out", paste0('nwis_', site, '_data.csv'))
+    
+    these_data <- download_nwis_site_data(file_out,start_date,end_date,parameterCd)
+    write_csv(these_data, file = file_out)
```
Good call, I made a new target `p1_site_data_csv` for this intermediate csv

abnerbog committed:
Removed calling and passing of filepath

abnerbog committed:
Use dplyr method for xwalk and joining operations

abnerbog commented on 01_fetch/src/download_nws_data.R:
```diff
@@ -0,0 +1,21 @@
+                     select(nws_conversion, NWS, USGS), 
+                     by = c("site_no"="USGS"))
+  
+  for(i in sites$NWS){
+    
+    url <- paste0("https://api.water.noaa.gov/nwps/v1/gauges/",i)
+
+    response_data <- request(url) %>%
+      req_perform() %>%
+      resp_body_json()
+
+    sites$flood_stage[which(sites$NWS %in% i)] <- response_data$flood$categories$minor$stage
+    sites$stage_units[which(sites$NWS %in% i)] <- response_data$flood$stageUnits
+
+  }
```
This is definitely more easy to read. Thanks for this suggestion!

abnerbog committed:
Remove comments at top of each phase script

abnerbog committed:
Remove writing of sf data

abnerbog committed:
Modify phase structure (fetch/process/visualize)

abnerbog committed:
Missed lumping in this change to previous commit

abnerbog committed:
Remove file directory as input for non file targets

abnerbog commented on 02_read.R:
```diff
@@ -0,0 +1,21 @@
+# This file includes the target recipes for targets that clean and prepare data.
```
That makes sense! In the original USGS workflow the only target in the read phase was associated with reading the shapefile: https://github.com/DOI-USGS/hurricane-harvey/tree/main/scripts/read

abnerbog committed:
Remove placeholder file from all scripts folders

abnerbog committed:
Add html to leaflet target filename

abnerbog committed:
Add svg suffix to hydrograph target name

abnerbog committed:
Remove excess commands in leaflet map saving

abnerbog commented on 04_visualize/src/visualize_harvey_map.R:
```diff
@@ -0,0 +1,129 @@
+  # add custom legend for usgs gage sites
+  m <- add_gage_site_legend(m, color = usgs_gage_color, label = "USGS Gage Sites", position = "bottomright")
+  
+  saveWidget(m, file = basename(out_file), selfcontained = TRUE)
+  file.copy(basename(out_file), out_file, overwrite = TRUE)
+  file.remove(basename(out_file))
```
Yes! Great call, thanks

abnerbog commented on 04_visualize/src/visualize_hydrographs.R:
```diff
@@ -0,0 +1,37 @@
+    ) +
+    scale_x_date(date_labels = "%d-%b-%Y") +
+    # Use a minimal theme for a clean look
+    theme_minimal() +
```
Looks great! Here is the updated plot:
<img width="1243" height="896" alt="image" src="https://github.com/user-attachments/assets/84a4feaa-0481-4d95-b7c9-482ad7dc9e43" />

abnerbog committed:
Use black/white theme for plotting hydrographs

abnerbog committed:
Represent extracted files from zip as file target

abnerbog commented on 03_process.R:
```diff
@@ -0,0 +1,21 @@
+
+p3 <- list(
+  tar_target(
+    p3_harvey_pts,
+    load_sf_data(p1_harvey_best_track_id,
+                 start_date,
+                 end_date,
+                 in_folder="02_read/out",
+                 out_file="03_process/out/harvey_tracking.csv")
```
Thanks for this Lindsay, fixed the errors and the pipeline should run on your end now!

abnerbog committed:
Remove uncessary additions to gitignore

abnerbog commented on .gitignore:
```diff
@@ -1,32 +1,29 @@
 # Ignore targets build artifacts
-_targets/*
+_targets/
+# Ignore gemini generated files
+*/src/gemini_generated_files/*
```
Hmmm -- I thought this was a leaflet artifact but I can't remember to be honest...just ended up removing this :)

lindsayplatt commented on .gitignore:
```diff
@@ -1,32 +1,29 @@
 */out/*
 !*.placeholder
+*/tmp/*
+!*.placeholder
```
Just to clarify - I'm suggesting removing the duplicate `!*.placeholder` (which now appears 2x in the `.gitignore` file), not doing anything to the `tmp` or `out` folders

lindsayplatt commented:
I am able to build this now! Since this is supposed to be shared broadly, I still would recommend adding Roxygen-style documentation for the functions (at least a title/description). Let's talk about that at our check-in.

abnerbog committed:
Create images folder and move viznetwork image

abnerbog committed:
Create `install.R` file with pkg dependencies

abnerbog committed:
Create `runtime.txt` file with R version

abnerbog committed:
Create notebook file

abnerbog committed:
Replace Python cells in notebook to R cells

abnerbog committed:
Add `Rglpk` to `install.R`

abnerbog committed:
Add `libglpk40` to `apt.txt

abnerbog committed:
Update pkg installation instructions

abnerbog committed:
Add Binder badge to README

abnerbog committed:
Add example roxygen documentation

abnerbog committed:
Add `libudunits2-dev` to `apt.txt`

abnerbog committed:
Use `environment.yml` instead

abnerbog committed:
Remove `apt.txt`

abnerbog commented:
@lindsayplatt it seems like the error launching the repo contents with Binder is being seen by others in the community. I added a comment in this issue (https://github.com/jupyterhub/mybinder.org-deploy/issues/3387#issuecomment-3220771487) so hopefully a fix will be recommended soon.

abnerbog committed:
Add pkg to env and specify a few required pkg versions

abnerbog committed:
Incorporate dataRetrieval updates from USGS water data API changes

abnerbog committed:
Use harvey sites vector as input to p1_site_info

abnerbog committed:
Updates to defining and handling gage site ids

abnerbog committed:
Minor changes to viz fxns to handle dataRetrieval fxn updates

abnerbog committed:
Add fxn documentation

abnerbog committed:
Add comments to main target script file

abnerbog committed:
Remove environment.yml and replace with install.R

abnerbog commented:
Hi @lindsayplatt, just re-requested a review here! Here is a summary of the changes:
- Updated workflow to include the new `dataRetrieval` functions in  [6423c22](https://github.com/CUAHSI/from-scripts-to-pipelines-workshop/pull/12/commits/6423c229edfb899a49e47fa6bf98973507b6ef09)
- Added comments and improved documentation in [073a573](https://github.com/CUAHSI/from-scripts-to-pipelines-workshop/pull/12/commits/073a5730dda52ec3e2f1ad0a155103d4d2dcd5d6) and [dbc20ed](https://github.com/CUAHSI/from-scripts-to-pipelines-workshop/pull/12/commits/dbc20ed36c5572602c8cc5f03d6aee1479a89e7c)

Once this is good to go I will port over the content to [a copy of the R notebook template](https://github.com/abnerbog/eds-notebook-submission-targets) from the EDS book publishers for rendering and making content changes to the notebook.

lindsayplatt commented:
@abnerbog thanks! Noted and added to my list to complete by tomorrow :)

lindsayplatt commented on 01_fetch.R:
```diff
@@ -45,7 +45,7 @@ p1 <- list(
   tar_target(
     p1_site_info,
-    download_nwis_site_info(p1_site_data)
+    download_nwis_site_info(harvey_sites)
```
The only problem here with using the original list rather than the unique sites returned with `p1_site_data` is that now the pipeline is not robust to a site missing data and therefore appearing in `harvey_sites` but not in `p1_site_data`. I think its fine to keep as-is, but just something to consider for building pipelines!

abnerbog replied:
Good point! Thanks for bringing up.

lindsayplatt reviewed (approved):
Looks good! That switch to the new fxns didn't seem too involved. I am leaving one small inline note about the pipeline robustness that does not need to be addressed but wanted to mention it. Lastly, it could be good to add a comment in `_targets.R` next to `dataRetrieval` that says you need at least[ v2.7.19](https://github.com/DOI-USGS/dataRetrieval/blob/main/NEWS#L16C15-L23) since this uses the functions for the modern API.

abnerbog commented:
Thanks Lindsay for the review!
